lib/pdn/pdn.c

```diff
@@ -147,30 +155,70 @@ static void on_cgev(const char *notif)
 			continue;
 		}
 
+		/* parse <cid> */
 		p += strlen(map[i].notif);
 		if (*p == ' ') {
-			cid = strtoul(p, &p, 10);
+			cid = (uint8_t)strtoul(p, &p, 10);
 		} else {
 			cid = CID_UNASSIGNED;
 		}
 
-		if (cid == pdn_act_notif.cid && map[i].event == PDN_EVENT_ACTIVATED) {
+		if (map[i].event == PDN_EVENT_ACTIVATED && cid == pdn_act_notif.cid) {
+			/* parse <reason> */
 			if (*p == ',') {
-				pdn_act_notif.reason = strtol(p + 1, NULL, 10);
+				pdn_act_notif.reason = (int8_t)strtol(p + 1, NULL, 10);
 			} else {
 				pdn_act_notif.reason = PDN_ACT_REASON_NONE;
 			}
 			k_sem_give(&sem_cgev);
 		}
 
 		SYS_SLIST_FOR_EACH_CONTAINER(&pdn_contexts, pdn, node) {
-			if ((pdn->context_id == cid || cid == CID_UNASSIGNED) && pdn->callback) {
+			if (pdn->callback && (pdn->context_id == cid || cid == CID_UNASSIGNED)) {
 				pdn->callback(pdn->context_id, map[i].event, 0);
 			}
 		}
+		return;
+	}
+}
 
+static void parse_cgev_apn_rate_ctrl(const char *notif)
+{
+	char *p;
+	int8_t cid;
+	struct pdn *pdn;
+	uint8_t apn_rate_ctrl_status;
+
+	/* +CGEV: APNRATECTRL STAT <cid>,<status>,[<time_remaining>] */
+	p = strstr(notif, "APNRATECTRL STAT");
+	if (!p) {
 		return;
 	}
+
+	/* parse <cid> */
+	p += strlen("APNRATECTRL STAT");
```

Comment:
Why is it acceptable for PDN to have `CID_UNASSIGNED` when parsing `<cid>` while it isn't for APN?

Reply:
See my previous comment https://github.com/nrfconnect/sdk-nrf/pull/12311#issuecomment-1717565735